src/util/ecdsa.rs

```diff
@@ -412,6 +416,72 @@ impl<'de> ::serde::Deserialize<'de> for PublicKey {
     }
 }
 
+/// An ECDSA signature with the corresponding hash type.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct EcdsaSig {
+    /// The underlying ECDSA Signature
+    pub sig: secp256k1::Signature,
+    /// The corresponding hash type
+    pub hash_ty: LegacySigHashType,
+}
+
+impl EcdsaSig {
+
+    /// Deserialize from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, EcdsaSigError> {
+        match sl.split_last() {
+            Some((hash_ty, sig)) => {
+                let hash_ty = LegacySigHashType::from_u32_standard(*hash_ty as u32)
+                    .map_err(|_| EcdsaSigError::NonStandardSigHashType(*hash_ty))?;
+                let sig = secp256k1::Signature::from_der(sig)?;
+                Ok(EcdsaSig { sig, hash_ty })
+            },
+            None => {
+                Err(EcdsaSigError::Secp256k1(secp256k1::Error::InvalidSignature))?
+            },
+        }
+    }
+
+    /// Serialize EcdsaSig
+    pub fn serialize(&self) -> Vec<u8> {
```

Comment:
Shouldn't this be a `Encode` impl? I assume this is "consensus serialization". Also, why not to have `Decode` impl.